signer/src/context/mod.rs

```diff
@@ -58,14 +59,70 @@ pub trait Context: Clone + Sync + Send {
     /// later return `Some(_)`. But if [`StreamExt::next`] yields `None`
     /// three times then the stream is "fused" and will return `None`
     /// forever after.
-    fn as_signal_stream<M>(&self, network: &M) -> SelectAll<BroadcastStream<SignerSignal>>
+    fn as_signal_stream<F, M>(&self, network: &M, predicate: F) -> ReceiverStream<SignerSignal>
     where
         M: MessageTransfer,
+        F: Fn(&SignerSignal) -> bool + Send + Sync + 'static,
     {
-        let term = self.get_termination_handle().as_stream();
-        let signal_stream = BroadcastStream::new(self.get_signal_receiver());
-        let network_stream = network.receiver_stream();
+        let (sender, receiver) = tokio::sync::mpsc::channel(SIGNER_CHANNEL_CAPACITY);
 
-        futures::stream::select_all([term, signal_stream, network_stream])
+        let mut term = self.get_termination_handle().as_receiver();
+        let mut signal_stream = self.get_signal_receiver();
+        let mut network_stream = network.as_receiver();
+        tokio::spawn(async move {
+            loop {
+                tokio::select! {
+                    item = term.recv() => {
+                        match item {
+                            Some(signal) => {
+                                // An error means that the channel has been
+                                // closed. This is most likely due to the
+                                // receiver being closed so we can bail.
+                                if let Err(_) = sender.send(signal).await {
+                                    break;
+                                }
+                            }
+                            None => break,
+                        }
+                    }
+                    item = signal_stream.recv() => {
+                        match item {
+                            Ok(signal) if predicate(&signal) => {
+                                // See comment above, we can bail.
+                                if let Err(_) = sender.send(signal).await {
+                                    break;
+                                }
+                            }
+                            Ok(_) => continue,
+                            Err(RecvError::Closed) => {
+                                tracing::warn!("internal signal stream closed");
+                                break;
+                            }
+                            Err(error @ RecvError::Lagged(_)) => {
+                                tracing::warn!(%error, "internal signal stream lagging");
+                                continue
+                            }
+                        }
+                    }
+                    item = network_stream.recv() => {
```

Comment:
Hm, is this section needed right now due to the fact that tests use their special network thing right now?

Reply:
Oh yeah I think you're right. Libp2p sends messages on the signal stream, and so do our tests.

Reply:
I think the crux is rather that the in-memory networks _don't_ publish `P2PMessageReceived` etc :grimacing:

Reply:
I think so. This catch might be the source of all issues. Since we're going through this stream now, I don't think that we'll have this issue anymore. But I'm only 30% sure that I've thought this through.